src/syntax/lex.rs

```diff
@@ -0,0 +1,348 @@
+
+lazy_static! {
+    static ref CYRILLIC_CHARS: BTreeSet<char> = {
+        "АБВГДЕЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯабвгдеёжзийклмнопрстуфхцчшщъыьэюя"
+            .chars()
+            .collect::<BTreeSet::<_>>()
+    };
+}
+
+fn is_cyrillic(ch: char) -> bool {
+    CYRILLIC_CHARS.contains(&ch)
+}
+
+pub struct Lexer<CharIter: FusedIterator<Item = char>> {
+    intern: Peekable<CharIter>,
+    token_stream: Vec<Token>,
+    diags: Vec<Diagnostic>,
+    line: u32,
+    column: u16
+}
+
+impl<CharIter: FusedIterator<Item = char>> Lexer<CharIter> {
+    pub fn new(iter: CharIter) -> Self {
+        Self {
+            intern: iter.peekable(),
+            token_stream: vec![],
+            diags: vec![],
+            line: 1,
+            column: 1
+        }
+    }
+
+    fn next_char(&mut self) -> Option<char> {
+        self.intern.next().map(|ch| {
+            if ch == '\n' {
+                self.line += 1;
+                self.column = 1;
+            }
+            ch
+        })
+    }
+
+    fn peek_char(&mut self) -> Option<char> {
+        self.intern.peek().map(|x| *x)
+    }
+
+    fn cur_char_pos(&self) -> SourceRange {
+        SourceRange::new(self.line, self.column, self.column + 1)
+    }
+
+    fn skip_whitespace(&mut self) {
+        while let Some(ch) = self.peek_char() {
+            match ch {
+                ' ' | '\t' | '\r' | '\n' => {
+                    self.next_char();
+                },
+                _ => {
+                    break;
+                }
+            }
+        }
+    }
+
+    fn lex_keyword_or_magic_sym(&mut self) {
+        debug_assert!(self.peek_char().unwrap().is_alphanumeric()
+            || is_cyrillic(self.peek_char().unwrap()));
+
+        let start_col = self.column;
+        let mut s = String::new();
+        while let Some(ch) = self.peek_char() {
+            if ch.is_alphanumeric() || is_cyrillic(ch) || ch == '!' || ch == '?' {
+                self.next_char();
+                s.push(ch);
+            } else {
+                break;
+            }
+        }
+
+        self.token_stream.push(Token::new(
+            if let Some(token_kind) = lookup_keyword(&s) {
+                token_kind
+            } else {
+                TokenKind::Identifier(s)
+            },
+            SourceRange::new(self.line, start_col, self.column)));
+    }
+
+    fn lex_number(&mut self) {
+        debug_assert!(self.peek_char().unwrap().is_numeric());
+
+        let start_col = self.column;
+        let mut num = 0i64;
+        while let Some(ch) = self.peek_char() {
+            if ch.to_digit(10).map(|digit| {
+                self.next_char();
+                num *= 10;
+                num += digit as i64;
+            }).is_none() {
+                break;
+            }
+        }
+        self.token_stream.push(Token::new(
+            TokenKind::LitNumber(num),
+            SourceRange::new(self.line, start_col, self.column)
+        ));
+    }
+
+    fn lex_char(&mut self) {
```

Comment:
the escape character of HNM is '?', such as "?n", "?t", "??"

Reply:
fixed in new commit.